src/test/unit/io/json/json_data_tuples_test.cpp

```diff
@@ -0,0 +1,163 @@
+#include <stan/io/json/json_data.hpp>
+#include <stan/io/json/json_data_handler.hpp>
+#include <stan/io/json/json_error.hpp>
+#include <stan/io/json/json_handler.hpp>
+#include <stan/io/json/rapidjson_parser.hpp>
+
+#include <test/unit/io/json/util.hpp>
+
+#include <boost/limits.hpp>
+#include <boost/math/concepts/real_concept.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <gtest/gtest.h>
+#include <complex>
+#include <iostream>
+#include <fstream>
+
+// (int, real) x;
+TEST(ioJson, jsonData_tuple_int_real) {
+  std::vector<std::string> json_path;
+  json_path = {"src", "test", "test-data", "tuple_int_real.json"};
+  std::string filename = paths_to_fname(json_path);
+  std::ifstream in(filename);
+  stan::json::json_data jdata(in);
+  std::vector<int> expected_vals_i;
+  std::vector<double> expected_vals_r;
+  std::vector<size_t> expected_dims;
+  expected_vals_i.push_back(1);
+  test_int_var(jdata, "x.1", expected_vals_i, expected_dims);
+  expected_vals_r.push_back(6.28);
+  test_real_var(jdata, "x.2", expected_vals_r, expected_dims);
+  expected_vals_r.clear();
+  expected_vals_r.push_back(0.0000123);
+  test_real_var(jdata, "y", expected_vals_r, expected_dims);
+}
+
+// (real, (int, real)) x;
+TEST(ioJson, jsonData_tuple_nested) {
+  std::vector<std::string> json_path;
+  json_path = {"src", "test", "test-data", "tuple_nested.json"};
+  std::string filename = paths_to_fname(json_path);
+  std::ifstream in(filename);
+  stan::json::json_data jdata(in);
+  std::vector<int> expected_vals_i;
+  std::vector<double> expected_vals_r;
+  std::vector<size_t> expected_dims;
+  expected_vals_r.push_back(3.214);
+  test_real_var(jdata, "x.1", expected_vals_r, expected_dims);
+  expected_vals_i.push_back(1);
+  test_int_var(jdata, "x.2.1", expected_vals_i, expected_dims);
+  expected_vals_r.clear();
+  expected_vals_r.push_back(6.28);
+  test_real_var(jdata, "x.2.2", expected_vals_r, expected_dims);
+  expected_vals_r.clear();
+  expected_vals_r.push_back(0.0000123);
+  test_real_var(jdata, "y", expected_vals_r, expected_dims);
+}
+
+// array[2] (array[3] real, real) x;
```

Comment:
It seems like this comment is out of date. I like having them here (I find it makes the tests much more readable) but it will be difficult to enforce that they're correct.

Reply:
agreed.  JSON doesn't allow comments and Stan doesn't allow string variables, so no way to put them closer to the JSON data.   will remove.